### PR TITLE
doc: Add to Doxygen documentation guidelines

### PR DESCRIPTION
**Pull request by ch4ot1c**
Add some tips to developer-notes.md, gathered while doing #16947. Also suggests the simpler `:` notation for Javadoc arguments and adds `@return` and `@param[out]` example.

### CODE REVIEW

File: doc/developer-notes.md [hunk 1]
```diff
@@ -189,9 +191,18 @@ Not OK (used plenty in the current source, but not picked up):
 A full list of comment syntaxes picked up by Doxygen can be found at http://www.doxygen.nl/manual/docblocks.html,
 but the above styles are favored.
 
+Other recommendations:
+
+- Avoid redundant type and in/out information in function descriptions
```
Comment (laanwj) [inline doc/developer-notes.md]: Can you give an example for all of these? (like the cases above)

File: doc/developer-notes.md [hunk 2]
```diff
@@ -189,9 +191,18 @@ Not OK (used plenty in the current source, but not picked up):
...
+- Use backticks (&#96;&#96;) to refer to `argument`s in function descriptions
```
Comment (jonatack) [inline doc/developer-notes.md]: suggestion for readability: s/`argument`s/`argument` names/

File: doc/developer-notes.md [hunk 3]
```diff
@@ -189,9 +191,18 @@ Not OK (used plenty in the current source, but not picked up):
...
+- Backticks aren't required when referring to a Function() that Doxygen already knows about; it will build a hyperlink automatically
+- Avoid linking to external documentation; links can break
+- `STRIP_CODE_COMMENTS = YES` in Doxyfile.in, so if you want a standalone comment to be preserved in the Doxygen-generated source code, it must either not conform to Javadoc or be qualified (e.g. begin with `\file`, etc).
```
Comment (jonatack) [inline doc/developer-notes.md]: End all or none of these item sentences with a period/full stop, e.g. "."; in this document the other item lists use periods/full stops.

Comment (jonatack) [reply doc/developer-notes.md]: Wrapping lines at 80 or 120 characters for these changes could be nice for source code readability on smaller screens.

**Review by jonatack — COMMENTED** [review]
A few suggestions below.

File: doc/developer-notes.md [hunk 4]
```diff
@@ -5,38 +5,42 @@ Developer Notes
...
-        - [Compiling for test coverage](#compiling-for-test-coverage)
-        - [Performance profiling with perf](#performance-profiling-with-perf)
-    - [Locking/mutex usage notes](#lockingmutex-usage-notes)
-    - [Threads](#threads)
-    - [Ignoring IDE/editor files](#ignoring-ideeditor-files)
+  - [Coding Style (General)](#coding-style-general)
+  - [Coding Style (C++)](#coding-style-c)
+  - [Coding Style (Python)](#coding-style-python)
+  - [Coding Style (Doxygen-compatible comments)](#coding-style-doxygen-compatible-comments)
+    - [Generating Documentation](#generating-documentation)
```
Comment (ch4ot1c) [inline doc/developer-notes.md]: VSCode plugin "Markdown All in One" updated the ToC automatically, oops - shall we keep?

File: doc/developer-notes.md (hunk 1 above)
Comment (ch4ot1c) [reply doc/developer-notes.md]: Added some examples and simplified recommendations.

File: doc/developer-notes.md (hunk 4 above)
Comment (laanwj) [reply doc/developer-notes.md]: TOC updates are welcome, but please keep the original spacing

**Comment by fanquake** [comment]
Closing "Up for Grabs". There's a couple of improvements here that could be picked up.

File: doc/developer-notes.md [hunk 5]
```diff
@@ -5,38 +5,42 @@ Developer Notes
...
+  - [Development tips and tricks](#development-tips-and-tricks)
+    - [Compiling for debugging](#compiling-for-debugging)
+    - [Compiling for gprof profiling](#compiling-for-gprof-profiling)
+    - [debug.log](#debuglog)
+    - [Testnet and Regtest modes](#testnet-and-regtest-modes)
+    - [DEBUG_LOCKORDER](#debuglockorder)
```
Comment (jonatack) [inline doc/developer-notes.md]: The #debug_lockorder link should not have been changed; the changed version no longer works.

**Review by jonatack — COMMENTED** [review]
Ok, will finish this up.

File: doc/developer-notes.md (hunk 5 above)
Comment (fanquake) [reply doc/developer-notes.md]: There are a couple improvements in the table of contents, but any unneeded formatting changes can be dropped.

Comment (jonatack) [reply doc/developer-notes.md]: Agreed. Minimising.

**Pull request closed without merging**